chain/sync_manager.go

```diff
@@ -0,0 +1,369 @@
+	workerChan     chan *types.TipSet
+}
+
+type syncResult struct {
+	ts      *types.TipSet
+	success bool
+}
+
+const syncWorkerCount = 3
+
+func NewSyncManager(sync SyncFunc) *SyncManager {
+	return &SyncManager{
+		bspThresh:       1,
+		peerHeads:       make(map[peer.ID]*types.TipSet),
+		syncTargets:     make(chan *types.TipSet),
+		syncResults:     make(chan *syncResult),
+		syncStates:      make([]*SyncerState, syncWorkerCount),
+		incomingTipSets: make(chan *types.TipSet),
+		activeSyncs:     make(map[types.TipSetKey]*types.TipSet),
+		doSync:          sync,
+		stop:            make(chan struct{}),
+	}
+}
+
+func (sm *SyncManager) Start() {
+	go sm.syncScheduler()
+	for i := 0; i < syncWorkerCount; i++ {
+		go sm.syncWorker(i)
+	}
+}
+
+func (sm *SyncManager) Stop() {
+	close(sm.stop)
+}
+
+func (sm *SyncManager) SetPeerHead(ctx context.Context, p peer.ID, ts *types.TipSet) {
+	log.Info("set peer head!", ts.Height(), ts.Cids())
+	sm.lk.Lock()
+	defer sm.lk.Unlock()
+	sm.peerHeads[p] = ts
+
+	if !sm.bootstrapped {
+		spc := sm.syncedPeerCount()
+		if spc >= sm.bspThresh {
+			// Its go time!
+			target, err := sm.selectSyncTarget()
+			if err != nil {
+				log.Error("failed to select sync target: ", err)
+				return
+			}
+
+			sm.incomingTipSets <- target
+			// TODO: is this the right place to say we're bootstrapped? probably want to wait until the sync finishes
+			sm.bootstrapped = true
+		}
+		log.Infof("sync bootstrap has %d peers", spc)
+		return
+	}
+
+	sm.incomingTipSets <- ts
+}
+
+type syncBucketSet struct {
+	buckets []*syncTargetBucket
+}
+
+func (sbs *syncBucketSet) Insert(ts *types.TipSet) {
+	for _, b := range sbs.buckets {
+		if b.sameChainAs(ts) {
+			b.add(ts)
+			return
+		}
+	}
+	sbs.buckets = append(sbs.buckets, &syncTargetBucket{
+		tips:  []*types.TipSet{ts},
+		count: 1,
+	})
+}
+
+func (sbs *syncBucketSet) Pop() *syncTargetBucket {
+	var bestBuck *syncTargetBucket
+	var bestTs *types.TipSet
+	for _, b := range sbs.buckets {
+		hts := b.heaviestTipSet()
+		if bestBuck == nil || bestTs.ParentWeight().LessThan(hts.ParentWeight()) {
+			bestBuck = b
+			bestTs = hts
+		}
+	}
+
+	sbs.removeBucket(bestBuck)
+
+	return bestBuck
+}
+
+func (sbs *syncBucketSet) removeBucket(toremove *syncTargetBucket) {
+	nbuckets := make([]*syncTargetBucket, 0, len(sbs.buckets)-1)
+	for _, b := range sbs.buckets {
+		if b != toremove {
+			nbuckets = append(nbuckets, b)
+		}
+	}
+	sbs.buckets = nbuckets
+}
+
+func (sbs *syncBucketSet) PopRelated(ts *types.TipSet) *syncTargetBucket {
+	for _, b := range sbs.buckets {
+		if b.sameChainAs(ts) {
+			sbs.removeBucket(b)
+			return b
+		}
+	}
+	return nil
+}
+
+func (sbs *syncBucketSet) Heaviest() *types.TipSet {
+	// TODO: should also consider factoring in number of peers represented by each bucket here
+	var bestTs *types.TipSet
+	for _, b := range sbs.buckets {
+		bhts := b.heaviestTipSet()
+		if bestTs == nil || bhts.ParentWeight().GreaterThan(bestTs.ParentWeight()) {
```

Comment:
This should probably not be the parent ts weight (but the one from `ChainStore.Weight`)

Reply:
You can't get the weight of a tipset without evaluating it, and in order to evaluate it, you have to sync it, and this is the logic that is deciding what to sync